commcare_connect/opportunity/helpers.py

```diff
@@ -0,0 +1,30 @@
+from django.db.models import BooleanField, Case, Count, DateTimeField, Max, Min, Q, When
+
+from commcare_connect.opportunity.models import Opportunity, OpportunityAccess
+
+
+def get_annotated_opportunity_access(opportunity: Opportunity):
+    access_objects = (
+        OpportunityAccess.objects.filter(opportunity=opportunity)
+        .select_related("user", "opportunityclaim")
+        .annotate(
+            last_visit_date_d=Max("user__uservisit__visit_date", filter=Q(user__uservisit__opportunity=opportunity)),
+            learn_progress_d=(Count("user__completed_modules") / Count("opportunity__learn_app__learn_modules")) * 100,
```

Comment:
I think the completed module count needs an opportunity filter, since right now it is counting all completed modules for the user.

Reply:
Additionally when I run this function locally, I get a division by zero error and I suspect it is here (since I think this is the only place that does division. My user has done zero learning. It would be good to test this at each stage of the opportunity lifecycle to ensure it can handle all states of data.

Reply:
> I think the completed module count needs an opportunity filter, since right now it is counting all completed modules for the user.
Thanks for pointing that out. I will add a fix for that and add a test which can check the data when opportunity is active with user having completed only some learning modules.

> Additionally when I run this function locally, I get a division by zero error and I suspect it is here (since I think this is the only place that does division. My user has done zero learning. It would be good to test this at each stage of the opportunity lifecycle to ensure it can handle all states of data.

I think this could be due to the Opportunity Learn App not having any learn modules as the division by zero error only comes when the divisor is 0 i.e learn_modules count in this case.

Reply:
Added extra test cases which cover no learn data, partial learn data, learn complete and assessment passed, and all data.